bftengine/src/bftengine/messages/PrePrepareMsg.cpp

```diff
@@ -77,11 +73,21 @@ void PrePrepareMsg::validate(const ReplicasInfo& repInfo) const {
 }
 
 PrePrepareMsg::PrePrepareMsg(ReplicaId sender, ViewNum v, SeqNum s, CommitPath firstPath, bool isNull, size_t size)
-    : MessageBase(
-          sender,
-          MsgCode::PrePrepare,
-          (((size + sizeof(PrePrepareMsgHeader)) < maxSizeOfPrePrepareMsg()) ? (size + sizeof(PrePrepareMsgHeader))
-                                                                             : maxSizeOfPrePrepareMsg()))
+    : PrePrepareMsg(sender, v, s, firstPath, "", isNull, size) {}
+
+PrePrepareMsg::PrePrepareMsg(ReplicaId sender,
+                             ViewNum v,
+                             SeqNum s,
+                             CommitPath firstPath,
+                             const std::string& spanContext,
+                             bool isNull,
+                             size_t size)
+    : MessageBase(sender,
+                  MsgCode::PrePrepare,
+                  spanContext.size(),
+                  (((size + sizeof(Header)) < maxMessageSize<PrePrepareMsg>())
+                       ? (size + sizeof(Header))
+                       : maxMessageSize<PrePrepareMsg>() - spanContext.size()))
```

Comment:
here you assume that `maxMessageSize()` includes the span context size, which may not be true in general...

Reply:
The idea is the following: if the payload size(`size`) + header size is less than the maximum size + span size, then use it.
In this case, the context' size is added to the given size in the `MessageBase`.
Otherwise, use the  `maximum size - span size` because of ^^.
I agree the memory management model is unclear and hard to follow, but I only tried to incorporate the new functionality into it.